File: Guanabara - Java/Meus exercícios POO/Aula09/src/main/java/encapsulamento/Livro.java
package encapsulamento;
public class Livro implements Publicacao { //a classe importará os métodos da interface
    //Declarando Atributos
    private String autor, titulo;
    private int totPag, pagAtual;
    private boolean aberto;
    private Pessoa leitor; //a variável `leitor` será uma instância da classe `Pessoa`
    
    //Método Construtor
    public Livro(String autor, String titulo, int totPag, Pessoa leitor) {
        //atributos definidos por parâmetros
        this.autor = autor;
        this.titulo = titulo;
        this.totPag = totPag;
        this.leitor = leitor;
        //atributos pré-definidos
        this.pagAtual = 0;
        this.aberto = false;
    }
    //Métodos Públicos Abstratos
    public String detalhes() { //esse método foi criado automaticamente: Alt+insert -> toString()
        return "Livro{" + "autor=" + autor + ", titulo=" + titulo + ",\ntotPag=" + totPag + ", pagAtual=" + pagAtual + //Note que retorna uma String
                ", aberto=" + aberto + ",\nleitor=" + leitor.getNome() + ", idade=" + leitor.getIdade() + ", sexo=" + leitor.getSexo()+'}';
    }
    //Métodos Acessores
    public String getAutor() {
        return autor;
    }

    public void setAutor(String autor) {
        this.autor = autor;
    }

    public String getTitulo() {
        return titulo;
    }

    public void setTitulo(String titulo) {
        this.titulo = titulo;
    }

    public int getTotPag() {
        return totPag;
    }

    public void setTotPag(int totPag) {
        this.totPag = totPag;
    }

    public int getPagAtual() {
        return pagAtual;
    }

    public void setPagAtual(int pagAtual) {
        this.pagAtual = pagAtual;
    }

    public boolean isAberto() {
        return aberto;
    }

    public void setAberto(boolean aberto) {
        this.aberto = aberto;
    }

    public Pessoa getLeitor() {
        return leitor;
    }

    public void setLeitor(Pessoa leitor) {
        this.leitor = leitor;
    }
    
    //Métodos da interface
    @Override
    public void abrir() {
        this.aberto = true;
    }

    @Override
    public void fechar() {
        this.aberto = false;
    }

    @Override
    public void folhear(int p) {
        if (p > this.totPag) {
            this.pagAtual = 0; 
        } else {
            this.pagAtual = p;
        }
    }

    @Override
    public void avancarPag() {
        this.pagAtual++;
    }

    @Override
    public void voltarPag() {
        this.pagAtual--;
    }
}

File: PHP/exercicios-PHPOO/Aula09-Livraria/Livro.php
<?php
require_once "Pessoa.php";
require_once 'Publicacao.php';
class Livro implements Publicacao
{
    //Atributos
    private $titulo;
    private $autor;
    private $totPaginas;
    private $pagAtual;
    private $aberto;
    private $leitor;    //esse leitor será uma pessoa, portanto importe a classe Pessoa.php
    //Métodos Especiais
    function detalhes()
    {
        echo "<hr>Livro: " . $this->getTitulo() . "<br>Escrito por: " . $this->getAutor() .
        "<br>Páginas: " . $this->getTotPaginas() . "<br>Página atual: " . $this->getPagAtual() . "<br>Sendo lido por: " . $this->leitor->getNome() . "<hr>"; //perceba a referência da referência (duplas setas ->) em [$this->leitor->getNome()] 
    }
    //Métodos Implementados da interface
    public function abrir()
    {
        $this->aberto = true;
    }
    public function fechar()
    {
        $this->aberto = false;
    }
    public function folhear($pagina)
    {
        if ($pagina > $this->totPaginas) {
            $this->pagAtual = 1; //se o usuário mandar folhear um nº página maior que o total do livro, retornará para o início (um)
        } else {
            $this->pagAtual = $pagina;
        }
    }
    public function avancarPag()
    {
        if ($this->pagAtual === $this->totPaginas) {
            echo "O livro já está na última página, não é possível avançar uma página";
        } else {
            $this->pagAtual++;
        }
    }
    public function voltarPag()
    {
        if ($this->pagAtual == 1) {
            echo "O livro ainda está na primeira página, não é possível voltar uma página";
        } else {
            $this->pagAtual--;
        }
    }
    //Método Construtor
    function __construct($titulo, $autor, $totPag, $leitor)
    {
        $this->titulo = $titulo;
        $this->autor = $autor;
        $this->totPaginas = $totPag;
        $this->leitor = $leitor;
        $this->aberto = false;
        $this->pagAtual = 0;
    }
    //Métodos Acessores
    function getTitulo()
    {
        return $this->titulo;
    }
    function getAutor()
    {
        return $this->autor;
    }
    function getTotPaginas()
    {
        return $this->totPaginas;
    }
    function getPagAtual()
    {
        return $this->pagAtual;
    }
    function getAberto()
    {
        return $this->aberto;
    }
    function getLeitor()
    {
        return $this->leitor;
    }
    function setTitulo($titulo)
    {
        $this->titulo = $titulo;
    }
    function setAutor($autor)
    {
        $this->autor = $autor;
    }
    function setTotPaginas($totPaginas)
    {
        $this->totPaginas = $totPaginas;
    }
    function setPagAtual($pagAtual)
    {
        $this->pagAtual = $pagAtual;
    }
    function setAberto($aberto)
    {
        $this->aberto = $aberto;
    }
    function setLeitor($leitor)
    {
        $this->leitor = $leitor;
    }
}
